Examples for cmdmessenger

adrienemery opened the pull request:
I've added examples to match 3 of the examples in the CmdMessenger Arduino library.  Ill keep adding more as I have time.  I also update the README. 

I removed the line `print "CMDS", self._commands` since when waiting for a ack it prints like crazy to the terminal.  I was thinking of maybe adding a `verbose` option to the `wait_for_ack` method.

Also used your advice and just used a timeout=0 in the serial object instead of separate threads - you were spot on about the callback thing in multiple threads. 

Finally I also have a branch on my fork called `pypi` that has all the setup files for pushing this module to PyPi so it could be installed with `pip`.  I can merge that in later if you want as it would make it easier for people to get up and running.

adrienemery committed:
Update gitignore

adrienemery committed:
add sendandrecieve.py

adrienemery committed:
Fail nicely

adrienemery committed:
refactoring

adrienemery committed:
Remove seperate serial thread

adrienemery committed:
Move everything into a class

adrienemery committed:
Workering sending and recieving multiple args

adrienemery committed:
Move examples into examples folder

adrienemery committed:
Update README

adrienemery committed:
Update documentation

Davideddu commented:
Good job, thank you!
I'm won't merge it yet, though, you need to fix a lot of things first.

Please change all the occurrences of "recieve" with "receive" throughout the repository. It's not an English word. I'll check the readme and the docs after I merge the PR, you don't have to worry about it.

Other things that I really don't like:
* the use of ON/OFF constants instead of True/False. This is not C, this is Python. True means "on", False means "off". True means 1, False means 0. You can even do math with them. Try True+True and see what happens.
* the use of classes to store the command names. Use lists, the way they work is closest to the way CmdMessenger uses enum on Arduino. If you want to add an item in the middle of the class you have to manually increase by one all the following. If you use a list you can simply add it in the middle and use its index as the command id. In any case, the commands class doesn't need to be an object.
* docs on a new line. Put them in the first line, right after `"""`, then put the closing `"""` on a new line by itself.
* `time.clock()`. No. Inaccurate. Use `time.time()`.
* `print "message"; sys.exit(1)`. Nope, don't do that. Error messages go to stderr. Either raise SystemExit or manually print to stderr (`print > sys.stderr, "message"`).
* the naming of your commits. You don't have to fix them this time, but ***please*** make sure next time you make them more descriptive. "refactoring", "fail nicely" don't tell anything. `git commit --amend` is your friend (until you've pushed the branch).

Please fix these things **with descriptive commit messages** so I can merge it.
Thank you!

adrienemery committed:
Fix the spelling of receive

adrienemery committed:
Fix the spelling of recieve in filenames

adrienemery committed:
Update doc-string formatting and use lists to store commands instead of a class.

adrienemery committed:
Allow True/False to represent 1/0 when sending data to Arduino.

True or False get converted to 1 or 0 respectively since the CmdMessenger Arduino library expects a 1 or 0 in its `readBoolArg` method.

adrienemery committed:
Change time.clock() to time.time()

adrienemery commented:
Thanks for your quick feedback. 
I addressed all the issues you had above, let me know if I missed anything.

Using a list is makes it much easier to maintaim commands :+1:

depau commented on cmdmessenger.py:
```diff
@@ -259,6 +258,10 @@ def send_cmd(self, msgid, *args, **kwargs):
+            if a is False:
+                a = 0
+            elif a is True:
+                a = 1
```
You don't need lines 261-264. Just use `self._file.write(str(int(a)))`. There is a very close relationship between boolean and integer, both are compatible: `bool(x)` is returns False if x is an integer and is == 0, or True if it's != 0, whereas int(x), float(x) return 0, 0.0 if x is a boolean and is False or 1, 1.0 if it's True.

Davideddu commented:
Awesome! Thank you very much!
Just fix the lines I commented in file ```cmdmessenger.py``` (remove 261-264 and use ```str(int(a))``` instead of ```str(a)``` in 265) and it'll be merged. I suggest *remove redundant boolean conversion in send_cmd* as the commit message :p

adrienemery commented:
good call - although that will block strings being able to be sent as arguments using `str(int(a))`.  What if I do:

```python
if isinstance(a, bool):
    a = int(a)
```

that way strings will stay as is.

Davideddu commented:
That works. You can also use ```type(a) == bool``` as the condition. Sorry I didn't realize other types could go in there.

adrienemery committed:
remove redundant boolean conversion in send_cmd

adrienemery commented:
done :)
